pennylane/operation.py

```diff
@@ -676,6 +676,85 @@ def __init__(self, *params, wires=None, do_queue=True):
         super().__init__(*params, wires=wires, do_queue=do_queue)
 
 
+class DiagonalOperation(Operation):
+    r"""Base class for diagonal quantum operations supported by a device.
+
+    As with :class:`~.Operation`, the following class attributes must be
+    defined for all operations:
+
+    * :attr:`~.Operator.num_params`
+    * :attr:`~.Operator.num_wires`
+    * :attr:`~.Operator.par_domain`
+
+    The following two class attributes are optional, but in most cases
+    should be clearly defined to avoid unexpected behavior during
+    differentiation.
+
+    * :attr:`~.Operation.grad_method`
+    * :attr:`~.Operation.grad_recipe`
+
+    Finally, there are some additional optional class attributes
+    that may be set, and used by certain quantum optimizers:
+
+    * :attr:`~.Operation.generator`
+
+    Args:
+        params (tuple[float, int, array, Variable]): operation parameters
+
+    Keyword Args:
+        wires (Sequence[int]): Subsystems it acts on. If not given, args[-1]
+            is interpreted as wires.
+        do_queue (bool): Indicates whether the operation should be
+            immediately pushed into a :class:`BaseQNode` circuit queue.
+            This flag is useful if there is some reason to run an Operation
+            outside of a BaseQNode context.
+    """
+    # pylint: disable=abstract-method
+
+    @abc.abstractstaticmethod
+    def _diagonal(*params):
+        """Array representation of the operator's diagonal
+        in the computational basis.
+
+        This is a *static method* that should be defined for all
+        new diagonal operations, that returns the array representing
+        the operator's diagonal in the computational basis.
+
+        This private method allows diagonals to be computed
+        directly without instantiating the operators first.
+
+        To return the diagonal of *instantiated* operators,
+        please use the :attr:`~.DiagonalOperator.diagonal` property instead.
+
+        **Example:**
+
+        >>> qml.RZ._diagonal(0.5)
+        array([0.96891242-0.24740396j, 0.96891242+0.24740396j])
+
+        Returns:
+            array: diagonal of the representation in the computational basis
+        """
+
+    @property
+    def diagonal(self):
+        """Return the diagonal of the operation.
+
+        Args:
+            abc ([type]): [description]
+        """
+        diag = self._diagonal(*self.parameters)
+
+        if self.inverse:
+            return diag.conj()
+
+        return diag
+
+    # TODO: make matrix a classmethod to enable simpler calculation
```

Comment:
Would `matrix` have to be made a classmethod on a higher level or here specifically?

Reply:
On the highest level where it is defined I'd say. But maybe it's sufficient to do it here. But I guess all the static methods should become classmethods.